vllm/sequence.py

```diff
@@ -1087,7 +1128,7 @@ class ExecuteModelRequest:
     # The number of requests in the running queue.
     running_queue_size: int = 0
     # Optional hidden states from prior step.
-    previous_hidden_states: Optional[HiddenStates] = None
+    previous_hidden_states: Union[HiddenStates, torch.Tensor, None] = None
```

Comment:
nit: push this complexity to `HiddenStates` definition:
`HiddenStates = Union[ActualHiddenStates, torch.Tensor]`
...
`previous_hidden_states: Optional[HiddenStates] = None`

need to think of a better name for `ActualHiddenStates`. I am not sure why it would be a torch tensor sometimes and a normal class in other cases.

Reply:
During the decode stage, the HiddenStates class stores only the hidden states of the last tokens that are propagated across steps. For the prefill stage we need hidden states of all tokens and there's no propagation across steps, so the hidden states are a simple tensor as we don't need any of the features implemented by HiddenStates class.

If it helps, we can rename HiddenStates class to DecodeHiddenStates and use it as following:
```
PrefillHiddenStates = torch.Tensor
HiddenStates = Union[PrefillHiddenStates, DecodeHiddenStates]
```

Otherwise, we can store prefill hidden states also using the same HiddenStates class and make seq_group_metadata_list optional (basically if seq_group_metadata_list is optional, we use it as a simple wrapper on tensor).

Reply:
The second solution you propose adds less complexity to vLLM, but the first works too. Will defer to you as it's a nit.

Reply:
States for the prefill step are now also handled by the HiddenStates class itself.